homie-controller/src/lib.rs

```diff
@@ -115,6 +115,7 @@ pub struct HomieController {
     /// The set of Homie devices which have been discovered so far, keyed by their IDs.
     // TODO: Consider using Mutex<im::HashMap<...>> instead.
     devices: Mutex<Arc<HashMap<String, Device>>>,
+    early_property_values: Mutex<Arc<HashMap<String, String>>>,
```

Comment:
The Mutex<Arc<_>> and .make_mut() dance is to support returning snapshots from the devices() method. I don't think you need to bother for early_property_values. We could probably document this better.

```suggestion
    /// The set of Homie devices which have been discovered so far, keyed by their IDs.
    /// This is Mutex<Arc<_>> rather than Arc<Mutex<_>> so that we can safely hand out
    /// snapshots in the public `devices()` method.
    // TODO: Consider using Mutex<im::HashMap<...>> instead.
    devices: Mutex<Arc<HashMap<String, Device>>>,
    /// Property values from messages that have arrived before the list of node properties is known.
    early_property_values: Mutex<HashMap<String, String>>,
```

Reply:
Ah, I see. Thanks for clearing that up.